packages/flutter/lib/src/services/asset_bundle.dart

```diff
@@ -91,30 +90,51 @@ class NetworkAssetBundle extends AssetBundle {
   /// to the given base URL.
   NetworkAssetBundle(Uri baseUrl)
     : _baseUrl = baseUrl,
-      _httpClient = createHttpClient();
+      _httpClient = new HttpClient();
 
   final Uri _baseUrl;
-  final http.Client _httpClient;
+  final HttpClient _httpClient;
 
-  String _urlFromKey(String key) => _baseUrl.resolve(key).toString();
+  Uri _urlFromKey(String key) => _baseUrl.resolve(key);
 
   @override
   Future<ByteData> load(String key) async {
-    final http.Response response = await _httpClient.get(_urlFromKey(key));
-    if (response.statusCode != 200)
-      throw new FlutterError('Unable to load asset: $key');
-    return response.bodyBytes.buffer.asByteData();
+    final HttpClientRequest request = await _httpClient.getUrl(_urlFromKey(key));
+    final HttpClientResponse response = await request.close();
+    if (response.statusCode != HttpStatus.OK)
+      throw new FlutterError(
+        'Unable to load asset: $key\n'
+        'HTTP status code: ${response.statusCode}'
+      );
+
+    final Completer<ByteData> completer = new Completer<ByteData>.sync();
+    if (response.contentLength == -1) {
+      final ByteConversionSink sink = new ByteConversionSink.withCallback((List<int> chunk) {
+        final Uint8List bytes = new Uint8List.fromList(chunk);
+        completer.complete(bytes.buffer.asByteData());
+      });
+      response.listen(sink.add, onDone: sink.close, onError: completer.completeError, cancelOnError: true);
+    } else {
+      final Uint8List bytes = new Uint8List(response.contentLength);
+      int offset = 0;
+      response.listen((List<int> chunk) {
+        for (int i = 0; i < chunk.length; i++, offset++) {
+          bytes[offset] = chunk[i];
+        }
+      },
+      onError: completer.completeError,
+      onDone: () {
+        completer.complete(bytes.buffer.asByteData());
+      },
+      cancelOnError: true);
+    }
```

Comment:
we should definitely not duplicate this code. see if you can factor it out into something in the foundation library, so that we only have to optimize it once.

Reply:
moved to `response_converter.dart` in foundation